### PR TITLE
Adding memory code

### PR DESCRIPTION
**Pull request by keefemitman**
Adding the functions that are necessary to compute the memory contributions to the strain as well as check the violation of the BMS balance laws.

### CODE REVIEW

File: scri/SpEC/__init__.py [hunk 1]
```diff
@@ -6,3 +6,5 @@
 
 from .com_motion import com_motion, estimate_avg_com_motion, remove_avg_com_motion
 from .file_io import read_from_h5, write_to_h5
+
+from .memory import mem_mass_aspect, mem_energy_flux, mem_angular_momentum_aspect, mem_angular_momentum_flux
```
Comment (10220) [inline scri/SpEC/__init__.py:10]: It seems like the main two functions in your `memory.py` script is `add_memory` and `BMS_strain`, did you mean to add those to this list?

File: scri/SpEC/memory.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/env python
```
Comment (10220) [inline scri/SpEC/memory.py:1]: So far this script just defines functions, so this line isn't needed at the current time.

File: scri/SpEC/memory.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+import numpy as np
+import spherical_functions as sf
+from spherical_functions import LM_index as lm
+from ..modes_time_series import ModesTimeSeries
+
+
+def WMtoMTS(h_wm):
```
[the remark this answers is not in the document]
Comment (keefemitman) [reply scri/SpEC/memory.py:9]: Got it, will do!

File: scri/SpEC/memory.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    if news == None:
+        news = h.copy()
+        news.data = h.data_dot
+    h_mts = WMtoMTS(h)
+    Psi2_mts = WMtoMTS(Psi2)
+    news_mts = WMtoMTS(news)
+
+    Bondi_M_mts = h_mts.copy()
+    Bondi_M_mts = (
+        0.5 * D_Op(-(Psi2_mts + 0.25 * 1.0j * h_mts.eth.eth.imag * -1.0j + 0.25 * (news_mts * h_mts.bar))).ethbar.ethbar
```
[the remark this answers is not in the document]
Comment (keefemitman) [reply scri/SpEC/memory.py:75]: Yes, all of my work in the memory paper follows the NP conventions. I checked the outputs of these functions as well to make sure everything matches up with the strain.

File: scri/SpEC/memory.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    )
+    # "* -1.0j" since we want the imaginary part, not the imaginary component
```
[the remark this answers is not in the document]
Comment (keefemitman) [reply scri/SpEC/memory.py:77]: Yes, from what I was finding ".imag" returns the imaginary component, not the imaginary part. I need just the imaginary part (i.e., the imaginary component without the factor of "i").

**Comment by 10220** [comment]
Hey @moble , after looking through Keefe's code it looks like there might be a bug (or a feature?) in `spherical_functions`: 

Let's I have a function f = 1 + 2i, then I decompose that into modes and store it as the `ModesTimeSeries` object `f_mts`. Then if I do `f_mts.imag`, this returns the modes corresponding to the function `2i` instead of `2`. Did you intend for this imaginary factor to remain present after taking the imaginary part?

**Comment by moble** [comment]
Evidently, I did make that decision at some point, because there's a test in `spherical_functions` that specifically requires that `f.imag.grid()` is a purely imaginary array.  But this was a terrible choice.  And if I'm not mistaken, [this line](https://github.com/moble/scri/blob/4ac1b624cfc8c7f7cc278d0bb241819ab907a332/scri/asymptotic_bondi_data/from_initial_values.py#L106) and [this line](https://github.com/moble/scri/blob/4ac1b624cfc8c7f7cc278d0bb241819ab907a332/scri/asymptotic_bondi_data/from_initial_values.py#L137) in `scri` both rely on that behavior (there would be a `1j * ` on each otherwise), but nothing else.  (The values returned by `constraint_mass_aspect` are both affected, but in the same way, so I don't see how that would change anything.  Please let me know if you see anything else.)

So, I'm going to make a new release of `spherical_functions` with the sensible behavior.  Then, I'll have to change those two lines in this package, require a sufficiently new version of `spherical_functions`, and make a new release of this package.  And then we can circle back to this PR.

Sorry about this.  I'll let you know when I'm done.

**Comment by moble** [comment]
Okay, the new versions are committed and have made it to pypi, so if you install either from source or with pip, update and you should be good to go.  I'm still working on the conda-forge release, but it should be available in a few hours...

**Comment by 10220** [comment]
Wow that was fast, thanks Mike!

**Comment by 10220** [comment]
This was implemented in the [sxs python package](https://github.com/sxs-collaboration/sxs). Can we close this PR?

**Comment by keefemitman** [comment]
Yes!

**Pull request closed without merging**